contrib/runners/winrm_runner/winrm_runner/winrm_base.py

```diff
@@ -224,22 +247,139 @@ def _translate_response(self, response):
         # objects so they can be used natively
         return (status, jsonify.json_loads(result, RESULT_KEYS_TO_TRANSFORM), None)
 
+    def _make_tmp_dir(self, parent):
+        LOG.debug("Creating temporary directory for WinRM script in parent: {}".format(parent))
+        ps = """$parent = {parent}
+$name = [System.IO.Path]::GetRandomFileName()
+$path = Join-Path $parent $name
+New-Item -ItemType Directory -Path $path | Out-Null
+$path""".format(parent=parent)
+        result = self._run_ps_or_raise(ps, ("Unable to make temporary directory for"
+                                            " powershell script"))
+        # strip to remove trailing newline and whitespace (if any)
+        return result['stdout'].strip()
+
+    def _rm_dir(self, directory):
+        ps = 'Remove-Item -Force -Recurse -Path "{}"'.format(directory)
+        self._run_ps_or_raise(ps, "Unable to remove temporary directory for powershell script")
+
+    def _upload(self, src_path_or_data, dst_path):
+        src_data = None
+        # detect if this is a path or a string containing data
+        # if this is a path, then read the data from the path
+        if os.path.exists(src_path_or_data):
+            LOG.debug("WinRM uploading local file: {}".format(src_path_or_data))
+            with open(src_path_or_data, 'r') as src_file:
+                src_data = src_file.read()
+        else:
+            LOG.debug("WinRM uploading data from a string")
+            src_data = src_path_or_data
+
+        # upload the data in chunks such that each chunk doesn't exceed the
+        # max command size of the windows command line
+        for i in range(0, len(src_data), WINRM_UPLOAD_CHUNK_SIZE_BYTES):
```

Comment:
Ah, I see.

So the script / file is still uploaded via command line string and not some other dedicated file upload protocol. There does appear to be overhead in serialization and de-serialization, but I imagine that's not an issue unless the script is very large (in 10's of megabytes or more).